### ISSUE TITLE
OIDC DCR not working with Keycloak

### ISSUE THREAD

**Issue by wkloucek** [owncloud/client#8816]
### Expected behaviour
1. see Steps to reproduce
4. The browser opens to let you log in - and uses the newly registered client ID via OIDC DCR
5. Keycloak show you the login dialog

### Actual behaviour
1. see Steps to reproduce
4. The browser opens to let you log in - but uses the static client ID which was not configured (xdXOt13JKxym1B1QcEncf2XDkLAexMBFwiT9j6EfhhHFJhs2KM9jbjTmf8JBXE69)  `https://keycloak.keycloak-test.owncloud.works/auth/realms/oCIS/protocol/openid-connect/auth?response_type=code&client_id=xdXOt13JKxym1B1QcEncf2XDkLAexMBFwiT9j6EfhhHFJhs2KM9jbjTmf8JBXE69&redirect_uri=http://localhost:41651&code_challenge=QmJccGatoH9JQvPIjDhEGdD1bJLge7Z19XhrXXLsZo0&code_challenge_method=S256&scope=openid%20offline_access%20email%20profile&prompt=select_account%20consent&state=3BrNuYqaa5CN-cLu5SBZU2T19y-kPLX6HY7hJ864bBU%3D`
5. Keycloak says `unkown client`

### Steps to reproduce
1. Setup [oCIS with Keycloak](https://github.com/owncloud/ocis/tree/master/deployments/examples/ocis_keycloak)
2. Do NOT set up [static client configuration for Desktop Client](https://github.com/owncloud/ocis/blob/master/deployments/examples/ocis_keycloak/config/keycloak/clients/desktop_client.json)
3. Use your Desktop client on that instance
4. The browser opens to let you log in - see actual / expected behaviour

### Server configuration
oCIS 1.9.0 with Keycloak

### Client configuration
Client version: 2.8.2
Operating system: Manjaro
OS language: English
Qt version used by client package: 5.15.2
Client package (From ownCloud or distro) (Linux only): distro

### Logs

1. Client logfile
<details>

```
07-14 14:13:53:092 [ debug gui.wizard ]	[ OCC::OwncloudWizard::slotCurrentPageChanged ]:	Current Wizard page changed to  0
07-14 14:14:02:838 [ debug gui.wizard ]	[ OCC::OwncloudSetupWizard::slotCheckServer ]:	Trying to look up system proxy
07-14 14:14:02:841 [ info gui.wizard ]:	No system proxy set by OS
07-14 14:14:02:842 [ info sync.accessmanager ]:	2 "" "https://ocis.keycloak-test.owncloud.works/status.php" has X-Request-ID "ae384266-8b2c-45a4-83e7-8f75bc86a9f0"
07-14 14:14:02:842 [ debug sync.cookiejar ]	[ OCC::CookieJar::cookiesForUrl ]:	QUrl("https://ocis.keycloak-test.owncloud.works/status.php") requests: ()
07-14 14:14:02:842 [ info sync.httplogger ]:	"ae384266-8b2c-45a4-83e7-8f75bc86a9f0: Request: GET https://ocis.keycloak-test.owncloud.works/status.php Header: { User-Agent: Mozilla/5.0 (Linux) mirall/2.8.2 (ownCloud, manjaro-5.13.1-3-MANJARO ClientArchitecture: x86_64 OsArchitecture: x86_64), Accept: */*, X-Request-ID: ae384266-8b2c-45a4-83e7-8f75bc86a9f0, } Data: []"
07-14 14:14:02:843 [ info sync.networkjob ]:	OCC::CheckServerJob created for "https://ocis.keycloak-test.owncloud.works" + "status.php" "OCC::OwncloudSetupWizard"
07-14 14:14:03:111 [ info sync.httplogger ]:	"ae384266-8b2c-45a4-83e7-8f75bc86a9f0: Response: GET 200 https://ocis.keycloak-test.owncloud.works/status.php Header: { Access-Control-Allow-Origin: *, Content-Length: 197, Content-Security-Policy: default-src 'none';, Content-Type: application/json, Date: Wed, 14 Jul 2021 12:14:03 GMT, Vary: Origin, X-Content-Type-Options: nosniff, X-Download-Options: noopen, X-Frame-Options: SAMEORIGIN, X-Permitted-Cross-Domain-Policies: none, X-Robots-Tag: none, X-Xss-Protection: 1; mode=block, } Data: [{\n    \"installed\": true,\n    \"maintenance\": false,\n    \"needsDbUpgrade\": false,\n    \"version\": \"10.0.9.5\",\n    \"versionstring\": \"10.0.9\",\n    \"edition\": \"community\",\n    \"productname\": \"ownCloud\"\n}]"
07-14 14:14:03:111 [ info sync.networkjob.checkserver ]:	status.php returns:  QJsonDocument({"edition":"community","installed":true,"maintenance":false,"needsDbUpgrade":false,"productname":"ownCloud","version":"10.0.9.5","versionstring":"10.0.9"})   QNetworkReply::NoError  Reply:  QNetworkReplyHttpImpl(0x5588000c0a30)
07-14 14:14:03:111 [ debug gui.wizard ]	[ OCC::OwncloudSetupWizard::slotFoundServer ]:	Successfully connected to "https://ocis.keycloak-test.owncloud.works" version "10.0.9" ( "10.0.9.5-ownCloud" )
07-14 14:14:03:111 [ info sync.networkjob.determineauthtype ]:	Determining auth type for QUrl("https://ocis.keycloak-test.owncloud.works/remote.php/webdav/")
07-14 14:14:03:111 [ info sync.accessmanager ]:	6 "PROPFIND" "https://ocis.keycloak-test.owncloud.works/remote.php/webdav/" has X-Request-ID "981bcf7a-7830-4126-a41e-b2ad9c85f28a"
07-14 14:14:03:111 [ debug sync.cookiejar ]	[ OCC::CookieJar::cookiesForUrl ]:	QUrl("https://ocis.keycloak-test.owncloud.works/remote.php/webdav/") requests: ()
07-14 14:14:03:111 [ info sync.httplogger ]:	"981bcf7a-7830-4126-a41e-b2ad9c85f28a: Request: PROPFIND https://ocis.keycloak-test.owncloud.works/remote.php/webdav/ Header: { User-Agent: Mozilla/5.0 (Linux) mirall/2.8.2 (ownCloud, manjaro-5.13.1-3-MANJARO ClientArchitecture: x86_64 OsArchitecture: x86_64), Accept: */*, Content-Type: text/xml; charset=utf-8, X-Request-ID: 981bcf7a-7830-4126-a41e-b2ad9c85f28a, } Data: []"
07-14 14:14:03:111 [ info sync.networkjob ]:	OCC::SimpleNetworkJob created for "https://ocis.keycloak-test.owncloud.works" + "" ""
07-14 14:14:03:111 [ debug sync.networkjob ]	[ OCC::AbstractNetworkJob::slotFinished ]:	Network job OCC::CheckServerJob finished for "status.php"
07-14 14:14:03:137 [ info sync.httplogger ]:	"981bcf7a-7830-4126-a41e-b2ad9c85f28a: Response: PROPFIND 401 https://ocis.keycloak-test.owncloud.works/remote.php/webdav/ Header: { Content-Length: 0, Date: Wed, 14 Jul 2021 12:14:03 GMT, Www-Authenticate: Basic realm=\"ocis.keycloak-test.owncloud.works\", Bearer realm=\"ocis.keycloak-test.owncloud.works\", } Data: []"
07-14 14:14:03:138 [ info sync.networkjob.determineauthtype ]:	Auth type for QUrl("https://ocis.keycloak-test.owncloud.works/remote.php/webdav/") is OCC::DetermineAuthTypeJob::AuthType::OAuth
07-14 14:14:03:138 [ info sync.accessmanager ]:	2 "" "https://ocis.keycloak-test.owncloud.works/.well-known/openid-configuration" has X-Request-ID "e2d20364-8e63-45aa-9895-d7533222ddd9"
07-14 14:14:03:138 [ debug sync.cookiejar ]	[ OCC::CookieJar::cookiesForUrl ]:	QUrl("https://ocis.keycloak-test.owncloud.works/.well-known/openid-configuration") requests: ()
07-14 14:14:03:139 [ info sync.httplogger ]:	"e2d20364-8e63-45aa-9895-d7533222ddd9: Request: GET https://ocis.keycloak-test.owncloud.works/.well-known/openid-configuration Header: { User-Agent: Mozilla/5.0 (Linux) mirall/2.8.2 (ownCloud, manjaro-5.13.1-3-MANJARO ClientArchitecture: x86_64 OsArchitecture: x86_64), Accept: */*, X-Request-ID: e2d20364-8e63-45aa-9895-d7533222ddd9, } Data: []"
07-14 14:14:03:139 [ info sync.networkjob ]:	OCC::SimpleNetworkJob created for "https://ocis.keycloak-test.owncloud.works" + "" ""
07-14 14:14:03:144 [ debug gui.wizard ]	[ OCC::OwncloudWizard::slotCurrentPageChanged ]:	Current Wizard page changed to  2
07-14 14:14:03:144 [ debug sync.networkjob ]	[ OCC::AbstractNetworkJob::slotFinished ]:	Network job OCC::SimpleNetworkJob finished for ""
07-14 14:14:03:253 [ info sync.httplogger ]:	"e2d20364-8e63-45aa-9895-d7533222ddd9: Response: GET 200 https://ocis.keycloak-test.owncloud.works/.well-known/openid-configuration Header: { Cache-Control: no-cache, must-revalidate, no-transform, no-store, Content-Length: 4058, Content-Type: application/json, Date: Wed, 14 Jul 2021 12:14:03 GMT, Referrer-Policy: no-referrer, Strict-Transport-Security: max-age=31536000; includeSubDomains, X-Content-Type-Options: nosniff, X-Frame-Options: SAMEORIGIN, X-Xss-Protection: 1; mode=block, } Data: [{\"issuer\":\"https://keycloak.keycloak-test.owncloud.works/auth/realms/oCIS\",\"authorization_endpoint\":\"https://keycloak.keycloak-test.owncloud.works/auth/realms/oCIS/protocol/openid-connect/auth\",\"token_endpoint\":\"https://keycloak.keycloak-test.owncloud.works/auth/realms/oCIS/protocol/openid-connect/token\",\"introspection_endpoint\":\"https://keycloak.keycloak-test.owncloud.works/auth/realms/oCIS/protocol/openid-connect/token/introspect\",\"userinfo_endpoint\":\"https://keycloak.keycloak-test.owncloud.works/auth/realms/oCIS/protocol/openid-connect/userinfo\",\"end_session_endpoint\":\"https://keycloak.keycloak-test.owncloud.works/auth/realms/oCIS/protocol/openid-connect/logout\",\"jwks_uri\":\"https://keycloak.keycloak-test.owncloud.works/auth/realms/oCIS/protocol/openid-connect/certs\",\"check_session_iframe\":\"https://keycloak.keycloak-test.owncloud.works/auth/realms/oCIS/protocol/openid-connect/login-status-iframe.html\",\"grant_types_supported\":[\"authorization_code\",\"implicit\",\"refresh_token\",\"password\",\"client_credentials\",\"urn:ietf:params:oauth:grant-type:device_code\",\"urn:openid:params:grant-type:ciba\"],\"response_types_supported\":[\"code\",\"none\",\"id_token\",\"token\",\"id_token token\",\"code id_token\",\"code token\",\"code id_token token\"],\"subject_types_supported\":[\"public\",\"pairwise\"],\"id_token_signing_alg_values_supported\":[\"PS384\",\"ES384\",\"RS384\",\"HS256\",\"HS512\",\"ES256\",\"RS256\",\"HS384\",\"ES512\",\"PS256\",\"PS512\",\"RS512\"],\"id_token_encryption_alg_values_supported\":[\"RSA-OAEP\",\"RSA-OAEP-256\",\"RSA1_5\"],\"id_token_encryption_enc_values_supported\":[\"A256GCM\",\"A192GCM\",\"A128GCM\",\"A128CBC-HS256\",\"A192CBC-HS384\",\"A256CBC-HS512\"],\"userinfo_signing_alg_values_supported\":[\"PS384\",\"ES384\",\"RS384\",\"HS256\",\"HS512\",\"ES256\",\"RS256\",\"HS384\",\"ES512\",\"PS256\",\"PS512\",\"RS512\",\"none\"],\"request_object_signing_alg_values_supported\":[\"PS384\",\"ES384\",\"RS384\",\"HS256\",\"HS512\",\"ES256\",\"RS256\",\"HS384\",\"ES512\",\"PS256\",\"PS512\",\"RS512\",\"none\"],\"response_modes_supported\":[\"query\",\"fragment\",\"form_post\"],\"registration_endpoint\":\"https://keycloak.keycloak-test.owncloud.works/auth/realms/oCIS/clients-registrations/openid-connect\",\"token_endpoint_auth_methods_supported\":[\"private_key_jwt\",\"client_secret_basic\",\"client_secret_post\",\"tls_client_auth\",\"client_secret_jwt\"],\"token_endpoint_auth_signing_alg_values_supported\":[\"PS384\",\"ES384\",\"RS384\",\"HS256\",\"HS512\",\"ES256\",\"RS256\",\"HS384\",\"ES512\",\"PS256\",\"PS512\",\"RS512\"],\"introspection_endpoint_auth_methods_supported\":[\"private_key_jwt\",\"client_secret_basic\",\"client_secret_post\",\"tls_client_auth\",\"client_secret_jwt\"],\"introspection_endpoint_auth_signing_alg_values_supported\":[\"PS384\",\"ES384\",\"RS384\",\"HS256\",\"HS512\",\"ES256\",\"RS256\",\"HS384\",\"ES512\",\"PS256\",\"PS512\",\"RS512\"],\"claims_supported\":[\"aud\",\"sub\",\"iss\",\"auth_time\",\"name\",\"given_name\",\"family_name\",\"preferred_username\",\"email\",\"acr\"],\"claim_types_supported\":[\"normal\"],\"claims_parameter_supported\":true,\"scopes_supported\":[\"openid\",\"email\",\"address\",\"offline_access\",\"profile\",\"phone\",\"roles\",\"web-origins\",\"microprofile-jwt\"],\"request_parameter_supported\":true,\"request_uri_parameter_supported\":true,\"require_request_uri_registration\":true,\"code_challenge_methods_supported\":[\"plain\",\"S256\"],\"tls_client_certificate_bound_access_tokens\":true,\"revocation_endpoint\":\"https://keycloak.keycloak-test.owncloud.works/auth/realms/oCIS/protocol/openid-connect/revoke\",\"revocation_endpoint_auth_methods_supported\":[\"private_key_jwt\",\"client_secret_basic\",\"client_secret_post\",\"tls_client_auth\",\"client_secret_jwt\"],\"revocation_endpoint_auth_signing_alg_values_supported\":[\"PS384\",\"ES384\",\"RS384\",\"HS256\",\"HS512\",\"ES256\",\"RS256\",\"HS384\",\"ES512\",\"PS256\",\"PS512\",\"RS512\"],\"backchannel_logout_supported\":true,\"backchannel_logout_session_supported\":true,\"device_authorization_endpoint\":\"https://keycloak.keycloak-test.owncloud.works/auth/realms/oCIS/protocol/openid-connect/auth/device\",\"backchannel_token_delivery_modes_supported\":[\"poll\"],\"backchannel_authentication_endpoint\":\"https://keycloak.keycloak-test.owncloud.works/auth/realms/oCIS/protocol/openid-connect/ext/ciba/auth\"}]"
07-14 14:14:03:267 [ debug sync.networkjob ]	[ OCC::AbstractNetworkJob::slotFinished ]:	Network job OCC::SimpleNetworkJob finished for ""
07-14 14:14:18:516 [ debug gui.wizard ]	[ OCC::OwncloudWizard::slotCurrentPageChanged ]:	Current Wizard page changed to  -1
```

</details>

**Comment by wkloucek** [comment]
@michaelstingl something to add? :thinking:

**Comment by michaelstingl** [comment]
@felix-schwarz please check

**Comment by jnweiger** [comment]
Retested with testpilocloud 2.9.0-beta3 against https://keycloak.ocis-keycloak.released.owncloud.works/auth/

* remove the static client_id from keycloak
* remove any trunsted hosts policy from keycloak

-> we confirm that a) dynamic registration was used (no fallback)
-> the ip address was used in the redirect_uri, as seen below:
 
http://127.0.0.1:40033/?state=SBGB-XRT8dFJvAc9lmZvTRaYgdrNglUDjFa5y88HgGw%3D&session_state=8281f3f6-ba4f-4d69-9fc5-fb9437913976&code=6526e72d-9c29-4190-b820-ef77d24f4ca1.8281f3f6-ba4f-4d69-9fc5-fb9437913976.0d4e531e-fde0-425d-908e-aea3242f64d1

Confirmed fixed in client 2.9.0
